package/MDAnalysis/core/groups.py

```diff
@@ -28,48 +43,81 @@ def copy_class(newname, *parents):
-    Segmentlevel = levels.Level('segment', S, SG)
-
-    Atomlevel.parent = Residuelevel
-    Atomlevel.child = None
-    Residuelevel.parent = Segmentlevel
-    Residuelevel.child = Atomlevel
-    Segmentlevel.parent = None
-    Segmentlevel.child = Residuelevel
-
-    A.level = Atomlevel
-    AG.level = Atomlevel
-    R.level = Residuelevel
-    RG.level = Residuelevel
-    S.level = Segmentlevel
-    SG.level = Segmentlevel
+        'SegmentBase', ComponentBase)
 
     return classdict
 
 
+class _MutableBase(object):
+    """
+    Base class that merges appropriate :class:`TopologyAttr` container classes.
+
+    Implements :attr:`__new__`. In it the instantiating class is merged with
+    the appropriate base from :attr:`Universe._classes`. The resulting class is
+    saved in the :attr:`Universe._class_cache` dictionary. The instantiating
+    class itself is used as the dictionary key, for simplicity in cache
+    retrieval.
+    
+    The resulting merged class is also cached using itself as the key, to
+    simplify class reuse. This is unneeded in practice because
+    :class:`GroupBase` and :class:`ComponentBase` shortcut their
+    :attr:`__new__` to :attr:`object.__new__`, skipping this process when
+    reusing an already-merged class.
+    """
+    # This signature must be kept in sync with the __init__ signature of
+    #  GroupBase and ComponentBase.
+    def __new__(cls, ix, u):
+        try:
+            # Kept terse so that cache lookup has minimal performance impact.
+            return object.__new__(u._class_cache[cls])
+        except KeyError:
+            if cls.order == 'component':
+                objtype = cls.level.name
+            else:
+                objtype = cls.level.name + cls.order
+            try:
+                # The base classes should have MRO priority, so that the
+                #  __new__ shortcut in Group/ComponentBase can kick in.
+                # OTOH, this prevents overriding TopologyAttr properties
+                # when subclassing Atom/Residue/SegmentGroups.
+                newcls = type(cls.__name__, (u._classes[objtype], cls), {})
+            except KeyError:
+                raise TypeError("Attempted to instantiate class '{}' but its "
+                                "level/order combination ('{}') is "
+                                "unsupported. Currently implemented "
+                                "combinations are: {}".format(cls.__name__,
+                                    objtype, str(sorted(u._classes.keys()))))
+            u._class_cache[cls] = u._class_cache[newcls] = newcls
+            return object.__new__(newcls)
+
 class GroupBase(object):
     """Base class from which a Universe's Group class is built.
 
     """
+    # Since this class will come first when mixed with AtomGroup,
+    #  setting __new__ like this will avoid having to go through the
+    #  cache lookup if the class is reused (as in ag.__class__(...)).
+    __new__ = object.__new__
```

Comment:
I'm simplifying this shortcut via `object.__new__`. As it stands, in `_MutableBase.__new__` `GroupBase` subclasses must be merged with higher MRO priority, so that their `__new__` overrides `_MutableBase.__new__`.
This makes it impossible for `AtomGroup` subclasses to override `TopologyAttrs`.
If I move this `__new__ = object.__new__` shortcut to its own placeholder class (`_ImmutableBase`, for lack of a better name) I only need to give priority to `_ImmutableBase` when merging, and `GroupBase` can be added as the least priority parent.